packages/node/src/integrations/localvariables.ts

```diff
@@ -4,14 +4,39 @@ import { LRUMap } from 'lru_map';
 
 import type { NodeClientOptions } from '../types';
 
+type Variables = Record<string, unknown>;
+type OnPauseEvent = InspectorNotification<Debugger.PausedEventDataType>;
 export interface DebugSession {
   /** Configures and connects to the debug session */
-  configureAndConnect(
-    onPause: (message: InspectorNotification<Debugger.PausedEventDataType>) => void,
-    captureAll: boolean,
-  ): void;
+  configureAndConnect(onPause: (message: OnPauseEvent, complete: () => void) => void, captureAll: boolean): void;
   /** Gets local variables for an objectId */
-  getLocalVariables(objectId: string): Promise<Record<string, unknown>>;
+  getLocalVariables(objectId: string, callback: (vars: Variables) => void): void;
+}
+
+type Next<T> = (result: T) => void;
+type Add<T> = (fn: Next<T>) => void;
+type CallbackWrapper<T> = { add: Add<T>; next: Next<T> };
+
+function createCallbackList<T>(complete: Next<T>): CallbackWrapper<T> {
+  // A collection of callbacks to be executed last to first
+  const callbacks: Next<T>[] = [complete];
+
+  function add(fn: Next<T>): void {
+    callbacks.push(fn);
+  }
+
+  function next(result: T): void {
+    const popped = callbacks.pop() || complete;
```

Comment:
Should we no-op if callbacks is empty? Perhaps give some kind of indicator of success (boolean return) from the `next` function?

Reply:
Since we add `complete` to the top of the callbacks stack, the only way `next` can be called when callbacks is empty is if you call `next` more than once per `add`ed closure or manage to call `next` in `complete` (which would be tricky to do since you have to supply `complete` before `next` is returned). The `|| callback` here is mainly to keep the types happy and does not get hit currently.

If we don't expect to hit this path, maybe it would be better to throw if `pop` returns undefined?